index.js

```diff
@@ -16,25 +17,48 @@ function fastifyWebsocket (fastify, opts, next) {
     errorHandler = opts.errorHandler
   }
 
-  const options = Object.assign({}, opts.options)
-  if (options.path) {
-    fastify.log.warn('ws server path option shouldn\'t be provided, use a route instead')
+  if (opts.options && opts.options.noServer) {
+    return next(new Error("fastify-websocket doesn't support the ws noServer option. If you want to create a websocket server detatched from fastify, use the ws library directly."))
   }
-  if (!options.server && !options.noServer) {
-    options.server = fastify.server
+
+  const wssOptions = Object.assign({ noServer: true }, opts.options)
+
+  if (wssOptions.path) {
+    fastify.log.warn('ws server path option shouldn\'t be provided, use a route instead')
   }
 
-  const wss = new WebSocket.Server(options)
-  wss.on('connection', handleRouting)
+  // We always handle the upgrading manually so we can dispatch through the fastify stack before upgrading
+  const websocketListenServer = wssOptions.server || fastify.server
+  delete wssOptions.server
```

Comment:
Can you add a comment inline for this delete? Maybe we should throw instead?

Reply:
It's actually supported to pass `server` to this library still, this is just doing the dance to extract the option and use the passed in server differently. We add our own `upgrade` listener to the server (which I imagine is almost always `fastify.server`), dispatch the request through fastify, and then eventually do `handleUpgrade` later. There was already a test that tests this. I tried to explain that with the comment above, does that not do it? 

If we passed the server to `new WebSocketServer`, it would add its own `upgrade` listener and upgrade before we had a chance to run hooks.

Reply:
Added a bit more comment